src/Cms/LazyCollection.php

```diff
@@ -0,0 +1,324 @@
+ * are loaded and initialized lazily when they are
+ * first used.
+ *
+ * @package   Kirby Cms
+ * @author    Lukas Bestle <lukas@getkirby.com>
+ * @link      https://getkirby.com
+ * @copyright Bastian Allgeier
+ * @license   https://getkirby.com/license
+ *
+ * @template TValue
+ * @extends \Kirby\Cms\Collection<TValue>
+ */
+abstract class LazyCollection extends Collection
+{
+	/**
+	 * Flag that tells whether hydration has been
+	 * completed for all collection elements;
+	 * this is used to increase performance
+	 */
+	protected bool $hydrated = false;
+
+	/**
+	 * Temporary auto-hydration whenever a collection
+	 * method is called; some methods may not need raw
+	 * access to all collection data, so performance
+	 * will be improved if methods call hydration
+	 * themselves only if they need it
+	 * @deprecated
+	 * @todo Remove this in v6
+	 */
+	public function __call(string $key, $arguments)
+	{
+		$this->hydrateAll();
+		return parent::__call($key, $arguments);
+	}
+
+	/**
+	 * Low-level getter for elements
+	 *
+	 * @return TValue|null
+	 */
+	public function __get(string $key)
+	{
+		$element = parent::__get($key);
+
+		// `$element === null` could mean "element does not exist"
+		// or "element found but not hydrated"
+		if ($element === null && array_key_exists($key, $this->data)) {
+			return $this->hydrateElement($key);
+		}
+
+		return $element;
+	}
+
+	/**
+	 * Returns the current element
+	 * @deprecated
+	 * @todo Remove in v6
+	 *
+	 * @return TValue
+	 */
+	public function current(): mixed
+	{
+		$current = parent::current();
+
+		// `$current === null` could mean "empty collection"
+		// or "element found but not hydrated"
+		if ($current === null && $key = $this->key()) {
+			return $this->hydrateElement($key);
+		}
+
+		return $current;
+	}
+
+	/**
+	 * Filters elements by one of the
+	 * predefined filter methods, by a
+	 * custom filter function or an array of filters
+	 */
+	public function filter(string|array|Closure $field, ...$args): static
+	{
+		// to filter through values, we need all values present
+		$this->hydrateAll();
+
+		return parent::filter($field, ...$args);
+	}
+
+	/**
+	 * Returns the first element
+	 *
+	 * @return TValue
+	 */
+	public function first()
+	{
+		$first = parent::first();
+
+		// `$first === null` could mean "empty collection"
+		// or "element found but not hydrated"
+		if ($first === null && $key = array_key_first($this->data)) {
+			return $this->hydrateElement($key);
+		}
+
+		return $first;
+	}
+
+	/**
+	 * Returns an iterator for the elements
+	 * @return \Iterator<TKey, TValue>
+	 */
+	public function getIterator(): Iterator
+	{
+		foreach ($this->data as $key => $value) {
+			if ($value === null) {
+				$value = $this->hydrateElement($key);
+			}
+
+			yield $key => $value;
+		}
+	}
+
+	/**
+	 * Ensures that all collection elements are loaded,
+	 * essentially converting the lazy collection into a
+	 * normal collection
+	 */
+	public function hydrateAll(): void
```

Comment:
`Suggestion:` hydrating the collection would imply all elements

```suggestion
	public function hydrate(): void
```

Reply:
I like that. To avoid merge conflicts (especially because PR 5 also introduces `initializeAll()` that shares the same naming change), I'd implement this in a PR 6. I have added it to my local todo list.

Reply:
#7843